packages/components/src/components/navigation/utils.ts

```diff
@@ -0,0 +1,26 @@
+import { AppURL } from '../..';
+import { Location } from "history";
+
+export const ON_LEAVE_DIRTY_STATE_MESSAGE = 'You have unsaved changes which will be lost. Are you sure you want to continue?';
+
+/**
+ * This function can be used as the callback for react-router's setRouteLeaveHook.  It should be preferred
+ * over a callback that simply returns the confirm message because with react-router v3.x, the URL route
+ * will have already been changed by the time this confirm is shown and will not be reset if the user does not confirm.
+ * If the user tries to click on the initial link again after canceling, nothing will happen because the URL
+ * in the browser will not change. (Issue 39633).  See also https://stackoverflow.com/questions/32841757/detecting-user-leaving-page-with-react-router
+ *
+ * @param currentLocation the location of the current page
+ */
+export function confirmLeaveWhenDirty(currentLocation: Location) {
+    const result = confirm(ON_LEAVE_DIRTY_STATE_MESSAGE);
+    if (result) {
+        // navigation confirmed
+        return true;
+    } else {
+        // navigation canceled, pushing the previous path
+        if (currentLocation)
```

Comment:
I'm curious why the manual parsing of the location is necessary? I think this is handled by `react-router`. The [docs for setRouteLeaveHook](https://github.com/ReactTraining/react-router/blob/v3/docs/API.md#setrouteleavehookroute-hook) indicate two distinct scenarios:

1. Normal app transition - Returning a `string` will prompt the user for a confirmation. Returning `false` will prevent the transition without prompt.
1. `beforeunload` event transition - In this case it must return a prompt message (a `string`) to prevent the transition.

Reply:
It appears that transition to <Link> can be prevented by returning false, but that's not the case for links defined with <a href>. for <a>, the url is changed, although page content not reloaded. Not sure if this is fixed with v4...